simson/common/data_transformations.py

```diff
@@ -1,61 +1,105 @@
     SigmoidalExtrapolation,
     ExponentialExtrapolation,
     WeightedProportionalExtrapolation,
 )
 
 
-def extrapolate_stock(
-    historic_stocks: fd.StockArray,
-    dims: fd.DimensionSet,
-    parameters: dict[str, fd.Parameter],
-    curve_strategy: str,
-    target_dim_letters=None,
-):
-    """Performs the per-capita transformation and the extrapolation."""
-
-    if target_dim_letters is None:
-        historic_dim_letters = historic_stocks.dims.letters
-        target_dim_letters = ("t",) + historic_dim_letters[1:]
-    else:
-        historic_dim_letters = ("h",) + target_dim_letters[1:]
-
-    # transform to per capita
-    pop = parameters["population"]
-    historic_pop = fd.FlodymArray(dims=dims[("h", "r")])
-    historic_gdppc = fd.FlodymArray(dims=dims[("h", "r")])
-    historic_stocks_pc = fd.FlodymArray(dims=dims[historic_dim_letters])
-    stocks_pc = fd.FlodymArray(dims=dims[target_dim_letters])
-    stocks = fd.FlodymArray(dims=dims[target_dim_letters])
-
-    historic_pop[...] = pop[{"t": dims["h"]}]
-    historic_gdppc[...] = parameters["gdppc"][{"t": dims["h"]}]
-    historic_stocks_pc[...] = historic_stocks / historic_pop
-
-    if curve_strategy == "GDP_regression":
-        gdp_regression(historic_stocks_pc.values, parameters["gdppc"].values, stocks_pc.values)
-    elif curve_strategy == "Exponential_GDP_regression":
-        gdp_regression(
-            historic_stocks_pc.values,
-            parameters["gdppc"].values,
-            stocks_pc.values,
-            fitting_function_type="exponential",
-        )
-    else:
-        raise RuntimeError(
-            f"Extrapolation strategy {curve_strategy} is not defined. "
-            f"It needs to be 'GDP_regression'."
+class StockExtrapolation:
+
+    def __init__(
+        self,
+        historic_stocks: fd.StockArray,
+        dims: fd.DimensionSet,
+        parameters: dict[str, fd.Parameter],
+        curve_strategy: str,
+        target_dim_letters=None,
+    ):
+        self.historic_stocks = historic_stocks
+        self.dims = dims
+        self.parameters = parameters
+        self.curve_strategy = curve_strategy
+        self.target_dim_letters = target_dim_letters
+        self.find_regression_strategy()
+        self.find_extrapolation_class()
+        self.n_fit_prms = self.extrapolation_class.n_prms()
+        self.extrapolate()
+
+    def find_regression_strategy(self) -> Callable:
+        """For now, only a regression based on GDP is implemented."""
+        self.regression_strategy = self.gdp_regression
+
+    def find_extrapolation_class(self) -> OneDimensionalExtrapolation:
+        if self.curve_strategy == "GDP_regression":
+            self.extrapolation_class = SigmoidalExtrapolation
+        elif self.curve_strategy == "Exponential_GDP_regression":
+            self.extrapolation_class = ExponentialExtrapolation
+        else:
+            raise ValueError('fitting_function_type must be either "sigmoid" or "exponential".')
```

Comment:
I would pass the class to the `StockExtrapolation` init directly here, and add a validator along the lines of  
```
extrapolation_classes = {cls.__name__: cls for cls in OneDimensionalExtrapolation.__subclasses__()}
if curve_strategy not in extrapolation_classes:
    raise ValueError("foo") 
return extrapolation_classes[curve_strategy]
```

Reply:
That would require, of course, passing strings that exactly match the class names (which avoids confusion), and it would be good to rename `curve_strategy` to something that includes a hint towards it being a class. `stock_extrapolation_class` or so.